algo/src/main/java/org/neo4j/graphalgo/impl/metaPathComputation/MetaPathPrecomputeHighDegreeNodes.java

```diff
@@ -0,0 +1,318 @@
+import java.io.*;
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import java.util.stream.Stream;
+
+
+//TODO test correctness!
+
+public class MetaPathPrecomputeHighDegreeNodes extends MetaPathComputation {
+
+    private Degrees degrees;
+    private HeavyGraph graph;
+    private ArrayGraphInterface arrayGraphInterface;
+    private ArrayList<Integer> metaPathsWeights;
+    private int metaPathLength;
+    //private ArrayList<HashSet<Integer>> initialInstances;
+    private int currentLabelId = 0;
+    private HashMap<Integer, HashMap<String, HashSet<Integer>>> duplicateFreeMetaPaths;
+    private PrintStream out;
+    private PrintStream debugOut;
+    private int printCount = 0;
+    private double estimatedCount;
+    private long startTime;
+    private HashMap<Integer, Integer> labelDictionary;
+    private int ratioHighDegreeNodes;
+    private List<Integer> maxDegreeNodes;
+
+
+    public MetaPathPrecomputeHighDegreeNodes(HeavyGraph graph, ArrayGraphInterface arrayGraphInterface, Degrees degrees, int metaPathLength, int ratioHighDegreeNodes) throws IOException {
+        this.graph = graph;
+        this.arrayGraphInterface = arrayGraphInterface;
+        this.metaPathsWeights = new ArrayList<>();
+        this.metaPathLength = metaPathLength;
+        //this.initialInstances = new ArrayList<>();
+        /*for (int i = 0; i < arrayGraphInterface.getAllLabels().size(); i++) {
+            this.initialInstances.add(new HashSet<>());
+        }*/
+        this.out = new PrintStream(new FileOutputStream("Precomputed_MetaPaths_HighDegree.txt"));//ends up in root/tests //or in dockerhome
+        this.debugOut = new PrintStream(new FileOutputStream("Precomputed_MetaPaths_HighDegree_Debug.txt"));
+        this.estimatedCount = Math.pow(arrayGraphInterface.getAllLabels().size(), metaPathLength + 1);
+        this.labelDictionary = new HashMap<>();
+        this.degrees = degrees;
+        this.ratioHighDegreeNodes = ratioHighDegreeNodes;
+        this.duplicateFreeMetaPaths = new HashMap<Integer, HashMap<String, HashSet<Integer>>>();
+    }
+
+    public Result compute() {
+        debugOut.println("started computation");
+        startTime = System.nanoTime();
+        maxDegreeNodes = getMaxDegreeNodes();
+        HashMap<Integer, HashMap<String, HashSet<Integer>>> finalMetaPaths = computeAllMetaPaths();
+        long endTime = System.nanoTime();
+
+        debugOut.println("calculation took: " + String.valueOf(endTime - startTime));
+        debugOut.println("actual amount of metaPaths: " + printCount);
+        debugOut.println("total time past: " + (endTime - startTime));
+        debugOut.println("finished computation");
+
+        debugOut.println(endTime - startTime);
+        duplicateFreeMetaPaths.forEach((a, b) -> {out.print(a + ":"); b.forEach((c, d) -> out.print(c + "=" + d.stream().map(Object::toString).collect(Collectors.joining(",")) + "-")); out.print("\n");});
```

Comment:
extract method

Reply:
avoid last "-"

Reply:
How to avoid it? If we would have to open the file for that and it is possible to solve it after/during the split operation I would prefer the last option, because I think writing to a file takes more time.